components/tikv_util/src/file_system/mod.rs

```diff
@@ -1,13 +1,139 @@
-// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::fs::{self, OpenOptions};
-use std::io::{self, ErrorKind, Read};
+mod file;
+mod rate_limiter;
+
+pub use file::{File, OpenOptions};
+pub use rate_limiter::{get_io_rate_limiter, set_io_rate_limiter, IORateLimiter};
+
+use std::cell::Cell;
+use std::fs;
+use std::io::{self, ErrorKind, Read, Write};
 use std::path::Path;
 use std::sync::{Arc, Mutex};
 
 use openssl::error::ErrorStack;
 use openssl::hash::{self, Hasher, MessageDigest};
 
+#[derive(Debug)]
+pub enum IOOp {
+    Read,
+    Write,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum IOType {
+    Other,
+    Read,
+    Write,
+    Coprocessor,
+    Flush,
+    Compaction,
+    Replication,
+    LoadBalance,
+    Import,
+    Export,
+}
+
+thread_local! {
+    static IO_TYPE: Cell<IOType> = Cell::new(IOType::Other)
+}
+
+fn set_io_type(new_io_type: IOType) {
+    IO_TYPE.with(|io_type| {
+        io_type.set(new_io_type);
+    });
+}
+
+fn get_io_type() -> IOType {
+    IO_TYPE.with(|io_type| io_type.get())
+}
+
+pub struct WithIOType {
+    previous_io_type: IOType,
+}
+
+impl WithIOType {
+    pub fn new(new_io_type: IOType) -> WithIOType {
+        let previous_io_type = get_io_type();
+        set_io_type(new_io_type);
+        WithIOType { previous_io_type }
+    }
+}
+
+impl Drop for WithIOType {
+    fn drop(&mut self) {
+        set_io_type(self.previous_io_type);
+    }
+}
+
+/// Indicates how large a buffer to pre-allocate before reading the entire file.
+fn initial_buffer_size(file: &File) -> usize {
+    // Allocate one extra byte so the buffer doesn't need to grow before the
+    // final `read` call at the end of the file.  Don't worry about `usize`
+    // overflow because reading will fail regardless in that case.
+    file.metadata().map(|m| m.len() as usize + 1).unwrap_or(0)
+}
+
+/// Write a slice as the entire contents of a file.
+pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> io::Result<()> {
```

Comment:
where are these helper functions used?

Reply:
They are equivalents to std::fs's module functions.